src/backend/distributed/utils/connection_management.c

```diff
@@ -0,0 +1,541 @@
+			 *
+			 * 1) Generally starting a connection might fail, after calling
+			 *    this function, so calling code needs to handle that anyway.
+			 * 2) This might be used in code that transparently handles
+			 *    connection failure.
+			 */
+		}
+
+		/* no connection available, done if a new connection isn't desirable */
+		if (!(flags & NEW_CONNECTION))
+		{
+			return NULL;
+		}
+	}
+
+	/*
+	 * Check whether we're right now allowed to open new connections.
+	 *
+	 * FIXME: This should be removed soon, once all connections go through
+	 * this API.
+	 */
+	if (XactModificationLevel > XACT_MODIFICATION_DATA)
+	{
+		ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+						errmsg("cannot open new connections after the first modification "
+							   "command within a transaction")));
+	}
+
+	/*
+	 * Either no caching desired, or no pre-established, non-claimed,
+	 * connection present. Initiate connection establishment.
+	 */
+	connection = StartConnectionEstablishment(&key);
+
+	oldContext = MemoryContextSwitchTo(ConnectionContext);
+	entry->connections = lappend(entry->connections, connection);
+	MemoryContextSwitchTo(oldContext);
+
+	if (flags & SESSION_LIFESPAN)
+	{
+		connection->sessionLifespan = true;
+	}
+
+	connection->activeInTransaction = true;
+
+	return connection;
+}
+
+
+/*
+ * Synchronously finish connection establishment of an individual connection.
+ *
+ * TODO: Replace with variant waiting for multiple connections.
+ */
+void
+FinishConnectionEstablishment(MultiConnection *connection)
+{
+	/*
+	 * Loop until connection is established, or failed (possibly just timed
+	 * out).
+	 */
+	while (true)
+	{
+		ConnStatusType status = PQstatus(connection->conn);
+		PostgresPollingStatusType pollmode;
+
+		if (status == CONNECTION_OK)
+		{
+			return;
+		}
+
+		/* FIXME: retries? */
+		if (status == CONNECTION_BAD)
+		{
+			return;
+		}
+
+		pollmode = PQconnectPoll(connection->conn);
+
+		/*
+		 * FIXME: Do we want to add transparent retry support here?
+		 */
+		if (pollmode == PGRES_POLLING_FAILED)
+		{
+			return;
+		}
+		else if (pollmode == PGRES_POLLING_OK)
+		{
+			return;
+		}
+		else
+		{
+			Assert(pollmode == PGRES_POLLING_WRITING ||
+				   pollmode == PGRES_POLLING_READING);
+		}
+
+		/* Loop, to handle poll() being interrupted by signals (EINTR) */
+		while (true)
+		{
+			struct pollfd pollFileDescriptor;
+			int pollResult = 0;
+
+			pollFileDescriptor.fd = PQsocket(connection->conn);
+			if (pollmode == PGRES_POLLING_READING)
+			{
+				pollFileDescriptor.events = POLLIN;
+			}
+			else
+			{
+				pollFileDescriptor.events = POLLOUT;
+			}
+			pollFileDescriptor.revents = 0;
+
+			pollResult = poll(&pollFileDescriptor, 1, CLIENT_CONNECT_TIMEOUT_SECONDS_INT);
+
+			if (pollResult == 0)
+			{
+				/* timeout exceeded */
+			}
+			else if (pollResult > 0)
+			{
+				/* IO possible, continue connection establishment */
+				break;
+			}
+			else if (pollResult != EINTR)
+			{
+				/* retrying, signal */
```

Comment:
The problem is that doing so reliably and portably essentially requires doing something like postgres' latches do, since poll/select aren't guaranteed to be interrupted by signals. In 9.6 this is trivially implementable for multiple connections because we/I introduced WaitEventSets which can wait both on latches and multiple ports at the same time (https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/include/storage/latch.h;hb=HEAD).

If we just use a single connection, we can just use WaitLatchOrSocket, but if we want to wait on multiple ones, we have to wait for 9.6 :(